### PR TITLE
Migrate to dskit/ring

### PR DESCRIPTION
**Pull request by treid314**
Signed-off-by: Tyler Reid <tyler.reid@grafana.com>

**What this PR does**: This PR migrates cortex to use dskit/ring instead of the local version. This could impact logging as we've moved the dskit ring package off of the global logger and used component loggers instead. 

**Checklist**
- [X] Tests updated
- [-] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/configuration/config-file-reference.md [hunk 1]
```diff
@@ -683,6 +683,11 @@ lifecycler:
     # CLI flag: -distributor.zone-awareness-enabled
     [zone_awareness_enabled: <boolean> | default = false]
 
+    # Comma-separated list of zones to exclude from the ring. Instances in
+    # excluded zones will be filtered out from the ring.
+    # CLI flag: -distributor.excluded-zones
+    [excluded_zones: <string> | default = ""]
```
Comment (pracucci) [inline docs/configuration/config-file-reference.md:689]: Please add this to CHANGELOG.

File: docs/configuration/config-file-reference.md [hunk 2]
```diff
@@ -701,10 +706,13 @@ lifecycler:
...
-  # Minimum duration to wait before becoming ready. This is to work around race
-  # conditions with ingesters exiting and updating the ring.
+  # Minimum duration to wait after the internal readiness checks have passed but
+  # before succeeding the readiness endpoint. This is used to slowdown
+  # deployment controllers (eg. Kubernetes) after an instance is ready and
+  # before they proceed with a rolling update, to give the rest of the cluster
+  # instances enough time to receive ring updates.
   # CLI flag: -ingester.min-ready-duration
-  [min_ready_duration: <duration> | default = 1m]
+  [min_ready_duration: <duration> | default = 15s]
```
Comment (pracucci) [inline docs/configuration/config-file-reference.md:715]: Please add this to CHANGELOG.

File: docs/configuration/config-file-reference.md [hunk 3]
```diff
@@ -729,6 +737,14 @@ lifecycler:
   # CLI flag: -ingester.unregister-on-shutdown
   [unregister_on_shutdown: <boolean> | default = true]
 
+  # When enabled the readiness probe succeeds only after all instances are
+  # ACTIVE and healthy in the ring, otherwise only the instance itself is
+  # checked. This option should be disabled if in your cluster multiple
+  # instances can be rolled out simultaneously, otherwise rolling updates may be
+  # slowed down.
+  # CLI flag: -ingester.readiness-check-ring-health
+  [readiness_check_ring_health: <boolean> | default = true]
```
Comment (pracucci) [inline docs/configuration/config-file-reference.md:746]: Please add this to CHANGELOG.

File: pkg/alertmanager/multitenant.go [hunk 4]
```diff
@@ -414,15 +414,11 @@ func createMultitenantAlertmanager(cfg *MultitenantAlertmanagerConfig, fallbackC
 			return nil, errors.Wrap(err, "failed to initialize Alertmanager's lifecycler")
 		}
 
-		am.ring, err = ring.NewWithStoreClientAndStrategy(am.cfg.ShardingRing.ToRingConfig(), RingNameForServer, RingKey, ringStore, ring.NewIgnoreUnhealthyInstancesReplicationStrategy())
```
Comment (pracucci) [inline pkg/alertmanager/multitenant.go:417]: Few lines above, the call to `ring.NewBasicLifecycler()` is not wrapping the Prometheus registry to add the `cortex_` prefix. I think it should be fixed.

**Review by pracucci — APPROVED** [review]
Changes LGTM, module few comments I would like to see addressed.

**Review by pracucci — APPROVED** [review]

**Review by pstibrany — APPROVED** [review]
Did a quick pass and changes look good to me.

**Pull request merged**